### PR TITLE
VCAI Fixes

### PR DESCRIPTION
**Pull request by alexvins**
Overall game speed (in headless mode, gui performance is out of scope) has been improved. AI pathfinding is still a bottleneck, even inlined version of graph initialization is slow. Pathfinder constructor is now 2 times faster itself called less often but still uses 20% of total CPU time (was 50%). Several smaller mostly bonus related bottlenecks has been fixed too.

### CODE REVIEW

File: AI/VCAI/Goals/Explore.cpp [hunk 1]
```diff
@@ -213,8 +196,7 @@ int Explore::howManyTilesWillBeDiscovered(
 			int3 npos = int3(x, y, pos.z);
 			if(cbp->isInTheMap(npos)
 				&& pos.dist2d(npos) - 0.5 < radious
-				&& !ts->fogOfWarMap[npos.x][npos.y][npos.z]
-				&& hasReachableNeighbor(npos, h, cbp, aip))
```
Comment (nullkiller) [inline AI/VCAI/Goals/Explore.cpp]: Why do you think it is wrong? It is dead end exploring cancellation.

Comment (alexvins) [reply AI/VCAI/Goals/Explore.cpp]: This function counts unexplored tiles within sight radius. `hasReachableNeighbor` check does not make sense.

Comment (nullkiller) [reply AI/VCAI/Goals/Explore.cpp]: It checks if unexplored tile has reachable neighbor an thus can be visited later. Otherwise there is much lower probability it will be useful. Even if it is not just a rock but a tile behind some border. The only way to reach it in such case is to cast air walking or dimension door. AI can not do both for now.

Comment (nullkiller) [reply AI/VCAI/Goals/Explore.cpp]: It might be an option to count potentially visitable tile as 1 and potentially unreachable as 0.1 or something. If it behaves badly somewhere lets look onto particular case.

File: AI/VCAI/Goals/Explore.h [hunk 2]
```diff
@@ -51,12 +51,11 @@ namespace Goals
 		TSubgoal explorationNewPoint(HeroPtr h) const;
 		TSubgoal explorationBestNeighbour(int3 hpos, int radius, HeroPtr h) const;
 		TSubgoal explorationScanRange(HeroPtr h, std::vector<int3> & range) const;
-		bool hasReachableNeighbor(const int3 &pos, HeroPtr hero, CCallback * cbp, VCAI * vcai) const;
 
 		void getVisibleNeighbours(
-			const std::vector<int3> & tiles, 
-			std::vector<int3> & out, 
-			CCallback * cbp, 
+			const std::vector<int3> & tiles,
```
Comment (nullkiller) [inline AI/VCAI/Goals/Explore.h]: Wrong line endings?

File: AI/VCAI/Goals/Explore.cpp (hunk 1 above)
Comment (alexvins) [reply AI/VCAI/Goals/Explore.cpp]: If tile is not visible yet it is not reachable too. Reachable neighbour check only works for border of FoW.

File: AI/VCAI/Goals/Explore.h (hunk 2 above)
Comment (alexvins) [reply AI/VCAI/Goals/Explore.h]: Spaces before line ending.

File: AI/VCAI/Goals/Explore.cpp (hunk 1 above)
Comment (nullkiller) [reply AI/VCAI/Goals/Explore.cpp]: Yes, it is an expected behavior. The idea is that AI should not get close to border because this way it explores behind it and thus explores unreachable area.

Comment (nullkiller) [reply AI/VCAI/Goals/Explore.cpp]: Does it hurt performance? We can try to check if there is any visitable tile but do not require it being reachable. This way we will cancel exploration of rocks. But it requires cb->getTile()

Comment (ArseniyShestakov) [reply AI/VCAI/Goals/Explore.cpp]: Basically all changes you see in this branch are about what Alex found in Valgrind's Callgrind output.

Comment (nullkiller) [reply AI/VCAI/Goals/Explore.cpp]: Just removing code is not always the best action to optimize things. I am afraid AI might start spending more turns on exploring thus being less productive. I have some ideas to optimize Explore logically by introducing the middle stage when we scan an area around hero instead of all map. Having such stage allows to disable this check in full scan. Also Explore can try to acquire kind of tile accessibility matrix which will be allocated and cached in PathfindingManager

File: AI/VCAI/VCAI.cpp [hunk 3]
```diff
@@ -1844,7 +1844,7 @@ bool VCAI::isAccessibleForHero(const int3 & pos, HeroPtr h, bool includeAllies)
 			}
 		}
 	}
-	return cb->getPathsInfo(h.get())->getPathInfo(pos)->reachable();
+	return ah->isTileAccessible(h, pos);
```
Comment (nullkiller) [inline AI/VCAI/VCAI.cpp]: This is not equivalent. PathfindingManager will return true if tile is accessible through a battle or any different way and reachable() returns true only if it is directly accessible. But in general it should be replaced.

File: AI/VCAI/Goals/Explore.cpp (hunk 1 above)
Comment (nullkiller) [reply AI/VCAI/Goals/Explore.cpp]: Comment out this function. I will try to find a better solution. We just need to reduce scan area somehow. And I have a few ideas.

Comment (alexvins) [reply AI/VCAI/Goals/Explore.cpp]: I can not remove `howManyTilesWillBeDiscovered` entirely. It is used in both `explorationScanRange` and `explorationBestNeighbour`

Comment (nullkiller) [reply AI/VCAI/Goals/Explore.cpp]: I mean hasReachableNeighbor

Comment (alexvins) [reply AI/VCAI/Goals/Explore.cpp]: done

Comment (nullkiller) [reply AI/VCAI/Goals/Explore.cpp]: Thanks! I will try to do something about it in the nearest future.

File: AI/VCAI/VCAI.cpp (hunk 3 above)
Comment (alexvins) [reply AI/VCAI/VCAI.cpp]: Is there an equivalent in AIPathfinder?

File: AI/VCAI/Goals/Explore.cpp (hunk 1 above)
Comment (DjWarmonger) [reply AI/VCAI/Goals/Explore.cpp]: Wait, this piece of code replaced old one, improving both AI performance and speed. I don't quite see how you could remove it entirely.

Comment (nullkiller) [reply AI/VCAI/Goals/Explore.cpp]: hasReachableNeighbor is for dead end cancellation. It improves exploration behavior but obviously not performance

File: AI/VCAI/VCAI.cpp (hunk 3 above)
Comment (nullkiller) [reply AI/VCAI/VCAI.cpp]: Btw VisitTile itself is an example of code which will defenitelly break if swithced to 
ah->isTileAccessibleForHero because it expects direct access

Comment (nullkiller) [reply AI/VCAI/VCAI.cpp]: Technically it is possible to implement equivalent function. AIPathfinder nodes are tiles where a hero has to do something like battle or spell cast. So they are not neighbors like in normal CPathfinder. In order to move between them we still need to call cb->getPathsInfo() but only at tryRealize(VisitTile) when we actually decided to move a hero.

So if AIPathfinder path has the length of 1 it is directly accessible. We can implement another version of isAccessibleForHero and check any path length to be 1. 

But this is not needed. We just need to avoid using VisitTile/VisitObj directly but use ah->howToVisitTile/Obj instead. If it returns an empty vector of goals the object is not accessible. So we need to replace all direct usages of VistTile/Obj with ah->howToVisitTile/Obj whenever it is used and remove all logic from VisitTile regarding to checking if tile is accessible and invoking ClearWayTo. But this is my opinion and it is a lot of work.

Comment (alexvins) [reply AI/VCAI/VCAI.cpp]: The actual problem is lack of paths cache in Client, single hero cache misses a lot (10 times more path recalculations that should be).

Comment (nullkiller) [reply AI/VCAI/VCAI.cpp]: Or you can do it yourself. The same way as it is already done but you need to check that any node also has no previousNeighbor and specialAction

File: AI/VCAI/Goals/Explore.cpp (hunk 1 above)
Comment (alexvins) [reply AI/VCAI/Goals/Explore.cpp]: What if there is a quest artefact hidden in the mountains (only accessible by the Door by design), will AI be able to find it with dead end cancellation?

File: AI/VCAI/VCAI.cpp (hunk 3 above)
Comment (alexvins) [reply AI/VCAI/VCAI.cpp]: I will do something anyway, still not decided what exactly, may be even both.

File: AI/VCAI/Goals/Explore.cpp (hunk 1 above)
Comment (nullkiller) [reply AI/VCAI/Goals/Explore.cpp]: Good question. Still I feel like dead end cancellation is needed but at the same time there should be a way to explore all possible tiles as low priority task.

Comment (alexvins) [reply AI/VCAI/Goals/Explore.cpp]: Lets add dead end check only to `explorationBestNeighbour`. So hero will efficiently continue exploring but start at any tile.

**Comment by alexvins** [comment]
Main bottlenecks so far:
* bonus caching, building "cache keys" OTF (f.e. `IBonusBearer::hasBonusOfType`). 
*  `CGHeroInstance::maxMovePoints` usage without supplying TurnInfo

File: AI/VCAI/Goals/Explore.cpp (hunk 1 above)
Comment (nullkiller) [reply AI/VCAI/Goals/Explore.cpp]: This does not work 100% nice. Sure it helps at certain extent. AI goes to dead end, cancels it and now he has nothing to explore because in front is dead end and behind is already explored. It enables full scan without dead-end cancellation and finds tiles in front so he goes and explores them because they are very close.

Comment (nullkiller) [reply AI/VCAI/Goals/Explore.cpp]: It is very tricky, this exploration. Maybe with middle stage where we scan in range around hero with dead end cancellation and only after we we enable to find a target near we run full scan with disabled dead end we can achieve better behavior.

Comment (nullkiller) [reply AI/VCAI/Goals/Explore.cpp]: Also as an option we can cancel exploration for a whole turn if it once yield no results. Also exploration sometimes triggered twice or even more doing exactly the same thing. Conquer, GatherArmy and CollectRes can produce exploration for instance. Explore decomposes to Explore(hero) instead of directly returning found goals. Also as an option we can put Explore to the list of basic goals and remove it from everywhere else thus allowing to decompose it only once per hero move. Also we can move hero not one tile at a time but 3 tiles or half of exploration radius thus reducing amount of decomposition cycles

Comment (nullkiller) [reply AI/VCAI/Goals/Explore.cpp]: Also might be an option to check if vcmieagles ai is set and disable explore at all

Comment (DjWarmonger) [reply AI/VCAI/Goals/Explore.cpp]: Can we just skip this line for this PR and test non-controversial performance optimizations?

Comment (nullkiller) [reply AI/VCAI/Goals/Explore.cpp]: Comment it out, already done I suppose and I will think about Explore optimization later. I expect I have to establish some profiler locally on Windows and will be able to see what is going on myself.

Comment (DjWarmonger) [reply AI/VCAI/Goals/Explore.cpp]: There's already heroesUnableToExplore container with logic exactly for that.
However, it's difficult to exclude exploration at all. If AI wants to explore, it means either it didn't explore entire map yet or map contains no necessary objectives at all.

Comment (nullkiller) [reply AI/VCAI/Goals/Explore.cpp]: Probably you are right, it should not take much time if there is no visible tiles. The worst case probably is 50-75% open map

File: lib/PathfinderUtil.h [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+#include "mapping/CMapDefines.h"
+#include "CGameState.h"
+
+namespace PathfinderUtil
+{
+	using FoW = std::vector<std::vector<std::vector<ui8> > >;
+	using ELayer = EPathfindingLayer;
+
+	template<EPathfindingLayer::EEPathfindingLayer layer>
```
[the remark this answers is not in the document]
Comment (alexvins) [reply lib/PathfinderUtil.h]: It is an optimization. There a `switch` on layer inside the function, because now layer is a constant inside it can be completely optimized out by compiler.

File: AI/VCAI/Pathfinding/AIPathfinderConfig.cpp [hunk 5]
```diff
@@ -76,12 +76,14 @@ namespace AIPathfinding
 
 		bool isAffordableBy(HeroPtr hero, const AIPathNode * source) const
 		{
+#ifdef VCMI_TRACE_PATHFINDER
```
[the remark this answers is not in the document]
Comment (ArseniyShestakov) [reply AI/VCAI/Pathfinding/AIPathfinderConfig.cpp]: I really dislike this one though...

File: AI/VCAI/Pathfinding/AINodeStorage.cpp [hunk 6]
```diff
@@ -10,15 +10,61 @@
...
+	int3 pos;
+	int3 sizes = gs->getMapSize();
+	const auto & fow = static_cast<const CGameInfoCallback *>(gs)->getPlayerTeam(hero->tempOwner)->fogOfWarMap;
+	const PlayerColor player = hero->tempOwner;
+
+	//make 200% sure that these are loop invariants (also a bit shorter code), let compiler do the rest(loop unswitching)
+	const bool useFlying = options.useFlying;
+	const bool useWaterWalking = options.useWaterWalking;
+
+	for(pos.x=0; pos.x < sizes.x; ++pos.x)
```
Comment (FeniksFire) [inline AI/VCAI/Pathfinding/AINodeStorage.cpp]: will be nice to overload some operators in int3 and replace 3 loops with something like that `for (int3 pos; pos<sizes; ++pos)`

File: AI/VCAI/FuzzyEngines.cpp [hunk 7]
```diff
@@ -50,23 +50,33 @@ armyStructure evaluateArmyStructure(const CArmedInstance * army)
 	double shootersStrenght = 0;
 	ui32 maxSpeed = 0;
 
+	static const CSelector selectorSHOOTER = Selector::type(Bonus::SHOOTER);
+	static const std::string keySHOOTER = "type_"+std::to_string((si32)Bonus::SHOOTER);
```
[the remark this answers is not in the document]
Comment (dydzio0614) [reply AI/VCAI/FuzzyEngines.cpp]: I somewhat agree, while we can have our own "vcmi standard library" stuff, there is no reason to introduce custom shortcuts for trivial stuff that force people to learn codebase-specific things for full understanding and make them use it.

File: lib/CPathfinder.cpp [hunk 8]
```diff
@@ -17,14 +17,57 @@
...
 
 bool canSeeObj(const CGObjectInstance * obj)
 {
 	/// Pathfinder should ignore placed events
 	return obj != nullptr && obj->ID != Obj::EVENT;
 }
 
+void NodeStorage::initialize(const PathfinderOptions & options, const CGameState * gs, const CGHeroInstance * hero)
+{
+	int3 pos;
```
Comment (FeniksFire) [inline lib/CPathfinder.cpp]: code dublication with `AINodeStorage::initialize`?

File: lib/logging/CLogger.cpp [hunk 9]
```diff
@@ -387,7 +389,8 @@ void CLogConsoleTarget::setColorMapping(const CColorMapping & colorMapping) { th
 CLogFileTarget::CLogFileTarget(boost::filesystem::path filePath, bool append)
 	: file(std::move(filePath), append ? std::ios_base::app : std::ios_base::out)
 {
-	formatter.setPattern("%d %l %n [%t] - %m");
+//	formatter.setPattern("%d %l %n [%t] - %m");
```
Comment (FeniksFire) [inline lib/logging/CLogger.cpp]: why this change?

Comment (alexvins) [reply lib/logging/CLogger.cpp]: VCMI write so much logs, so date formatting in too slow

File: lib/CPathfinder.h [hunk 10]
```diff
@@ -82,21 +117,23 @@ struct DLL_LINKAGE CPathsInfo
...
 
-	mutable boost::mutex pathMx;
-
 	const CGHeroInstance * hero;
 	int3 hpos;
 	int3 sizes;
 	boost::multi_array<CGPathNode, 4> nodes; //[w][h][level][layer]
 
-	CPathsInfo(const int3 & Sizes);
+	CPathsInfo(const int3 & Sizes, const CGHeroInstance * hero_);
```
[the remark this answers is not in the document]
Comment (alexvins) [reply lib/CPathfinder.h]: I do not like to have method/constructor parameters with same name as existing field.

File: AI/VCAI/Pathfinding/AINodeStorage.cpp [hunk 11]
```diff
@@ -10,15 +10,61 @@
...
+	{
+		for(pos.y=0; pos.y < sizes.y; ++pos.y)
+		{
+			for(pos.z=0; pos.z < sizes.z; ++pos.z)
+			{
+				const TerrainTile * tinfo = &gs->map->getTile(pos);
```
[the remark this answers is not in the document]
Comment (FeniksFire) [reply AI/VCAI/Pathfinding/AINodeStorage.cpp]: yep, far better than `tinfo`

File: AI/VCAI/Pathfinding/AINodeStorage.cpp (hunk 6 above)
Comment (alexvins) [reply AI/VCAI/Pathfinding/AINodeStorage.cpp]: I have no idea how to make such custom iterator fast, you have to add not trivial logic to `++` and compare all 3 coordinates each iteration.

File: AI/VCAI/FuzzyEngines.cpp [hunk 12]
```diff
@@ -50,23 +50,33 @@ armyStructure evaluateArmyStructure(const CArmedInstance * army)
...
+
+	static const CSelector selectorFLYING = Selector::type(Bonus::FLYING);
+	static const std::string keyFLYING = "type_"+std::to_string((si32)Bonus::FLYING);
+
+	static const CSelector selectorSTACKS_SPEED = Selector::type(Bonus::STACKS_SPEED);
+	static const std::string keySTACKS_SPEED = "type_"+std::to_string((si32)Bonus::STACKS_SPEED);
```
[the remark this answers is not in the document]
Comment (alexvins) [reply AI/VCAI/FuzzyEngines.cpp]: It is possible for  some similar simple cases, but generic solution is impossible (selector can be _any_ functor with `bool(const Bonus*)` signature)

File: AI/VCAI/FuzzyEngines.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexvins) [reply AI/VCAI/FuzzyEngines.cpp]: done

File: AI/VCAI/Pathfinding/AINodeStorage.cpp [hunk 13]
```diff
@@ -10,15 +10,61 @@
...
 
-AINodeStorage::~AINodeStorage()
+AINodeStorage::~AINodeStorage() = default;
+
+void AINodeStorage::initialize(const PathfinderOptions & options, const CGameState * gs, const CGHeroInstance * hero)
 {
+	//TODO: fix this code duplication with NodeStorage::initialize, problem is to keep `resetTile` inline
+
+	int3 pos;
+	int3 sizes = gs->getMapSize();
```
[the remark this answers is not in the document]
Comment (alexvins) [reply AI/VCAI/Pathfinding/AINodeStorage.cpp]: done

File: AI/VCAI/Pathfinding/AINodeStorage.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (alexvins) [reply AI/VCAI/Pathfinding/AINodeStorage.cpp]: done

File: AI/VCAI/Goals/Explore.cpp (hunk 1 above)
Comment (DjWarmonger) [reply AI/VCAI/Goals/Explore.cpp]: What is "commented out", the function call is still removed in this PR.

Comment (nullkiller) [reply AI/VCAI/Goals/Explore.cpp]: I have a few ideas for myself.
1. add a middle stage where AI scans 20x20 area around hero for a new point
2. scan perimeter of visible area before expanding it by sight radius and if it yields any result - return it. If all perimeter tiles are blocked expand and scan with deadend cancellation disabled.
3. might be worth to split AI pathfinder API onto one updatePaths(std:vector<HeroPtr>) and the rest is const. This way we can drop synchronization using mutex at all I think.

Comment (alexvins) [reply AI/VCAI/Goals/Explore.cpp]: Reverted. Any further improvements considered out of scope.

File: lib/CPathfinder.cpp (hunk 8 above)
Comment (alexvins) [reply lib/CPathfinder.cpp]: additional comment added

File: client/VCMI_client.cbp [hunk 14]
```diff
@@ -13,6 +13,7 @@
 				<Option object_output="../obj/Client/Debug/x86" />
 				<Option type="1" />
 				<Option compiler="gcc" />
+				<Option parameters="--testmap=Maps/Reclamation --spectate --spectate-skip-battle --spectate-ignore-hero" />
```
Comment (nullkiller) [inline client/VCMI_client.cbp]: Is it an intended change?

Comment (alexvins) [reply client/VCMI_client.cbp]: Yes it is.

Comment (alexvins) [reply client/VCMI_client.cbp]: I decided it is better to have default run parameters for debug to be automated run for now.

File: lib/CPathfinder.cpp [hunk 15]
```diff
@@ -193,16 +195,24 @@ void MovementCostRule::process(
 	const PathfinderConfig * pathfinderConfig,
 	CPathfinderHelper * pathfinderHelper) const
 {
-	int turnAtNextTile = destination.turn, moveAtNextTile = destination.movementLeft;
+	bool textTurn = false;
```
[the remark this answers is not in the document]
Comment (alexvins) [reply lib/CPathfinder.cpp]: fixed

File: lib/CPathfinder.cpp [hunk 16]
```diff
@@ -211,6 +221,12 @@ void MovementCostRule::process(
 		remains = pathfinderHelper->movementPointsAfterEmbark(moveAtNextTile, cost, destination.action - 1);
 	}
 
+	if(textTurn)
+		costAtNextTile = static_cast<float>(turnAtNextTile) +  static_cast<float>(cost) / maxMovePoints;
```
[the remark this answers is not in the document]
Comment (alexvins) [reply lib/CPathfinder.cpp]: Should be fixed now.

File: lib/CPathfinder.cpp [hunk 17]
```diff
@@ -193,16 +195,24 @@ void MovementCostRule::process(
 	const PathfinderConfig * pathfinderConfig,
 	CPathfinderHelper * pathfinderHelper) const
 {
-	int turnAtNextTile = destination.turn, moveAtNextTile = destination.movementLeft;
+	bool textTurn = false;
+	float costAtNextTile = destination.cost;
```
Comment (nullkiller) [inline lib/CPathfinder.cpp]: Where is destination.cost set?

File: lib/CPathfinder.cpp [hunk 18]
```diff
@@ -297,12 +313,16 @@ void CPathfinder::calculatePaths()
...
+		int movement = source.node->moveRemains;
+		uint8_t turn = source.node->turns;
+		float cost = source.node->cost;
+
 		hlp->updateTurnInfo(turn);
 		if(!movement)
 		{
 			hlp->updateTurnInfo(++turn);
 			movement = hlp->getMaxMovePoints(source.node->layer);
+			cost = static_cast<float>(turn);
```
Comment (nullkiller) [inline lib/CPathfinder.cpp]: Do we need this anymore?

Comment (alexvins) [reply lib/CPathfinder.cpp]: Already removed

File: lib/CPathfinder.cpp (hunk 17 above)
Comment (alexvins) [reply lib/CPathfinder.cpp]: In `CPathfinder::calculatePaths()` main loop
